README.rst
===============================
conda environment helper script
===============================


.. image:: https://img.shields.io/pypi/v/cenv_script.svg
        :target: https://pypi.python.org/pypi/cenv_script

.. image:: https://img.shields.io/travis/technic/cenv_script.svg
        :target: https://travis-ci.com/technic/cenv_script

.. image:: https://readthedocs.org/projects/cenv-script/badge/?version=latest
        :target: https://cenv-script.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Wrapper around conda that automatically persists changes to environment.yml
It is quite dummy but useful in simple case scenarios.


* Free software: MIT license
* Documentation: https://cenv-script.readthedocs.io.


Features
--------

* You run ``cenv-script install numpy``
* It calls ``conda install numpy``, checks for errors and then updates corresponding environment.yml

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.0.2 (2020-05-10)
------------------

* Preserve keys order in yaml output

0.0.1 (2020-05-08)
------------------

* Initial github commit


tests/__init__.py
"""Unit test package for cenv_script."""


tests/test_cenv.py
#!/usr/bin/env python

"""Tests for `cenv` package."""

import os
from pathlib import Path
from tempfile import TemporaryDirectory
from contextlib import contextmanager
import unittest
from click.testing import CliRunner

from cenv import cenv
from cenv import cli


@contextmanager
def working_directory(directory):
    owd = os.getcwd()
    try:
        os.chdir(directory)
        yield directory
    finally:
        os.chdir(owd)


class TestCenv(unittest.TestCase):
    """Tests for `cenv` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_env_file_exists(self):
        with TemporaryDirectory() as tmpdir:
            work_dir = os.path.join(tmpdir, "foo", "bar")
            os.makedirs(work_dir)
            env_file = os.path.join(tmpdir, "environment.yml")
            with open(env_file, "w") as f:
                f.write("\n")
            with working_directory(work_dir):
                self.assertEqual(
                    cenv.find_environment_file(), Path(env_file).resolve(),
                )

    def test_env_file_not_found(self):
        # env file must not be in tmp directory path or its ancestors!
        # otherwise test will not work as expected!
        with TemporaryDirectory() as tmpdir:
            with working_directory(tmpdir):
                with self.assertRaises(cenv.CondaEnvException):
                    cenv.find_environment_file()

    def test_command_line_interface(self):
        """Test the CLI."""
        runner = CliRunner()
        result = runner.invoke(cli.main)
        assert result.exit_code == 0
        assert "cenv.cli.main" in result.output
        help_result = runner.invoke(cli.main, ["--help"])
        assert help_result.exit_code == 0
        assert "--help  Show this message and exit." in help_result.output


tests/test_cenv_script.py
#!/usr/bin/env python

"""Tests for `cenv_script` package."""

import os
from pathlib import Path
from tempfile import TemporaryDirectory
from contextlib import contextmanager
import unittest
from click.testing import CliRunner

from cenv_script import cenv_script
from cenv_script import cli


@contextmanager
def working_directory(directory):
    owd = os.getcwd()
    try:
        os.chdir(directory)
        yield directory
    finally:
        os.chdir(owd)


class TestCenv(unittest.TestCase):
    """Tests for `cenv_script` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_env_file_exists(self):
        with TemporaryDirectory() as tmpdir:
            work_dir = os.path.join(tmpdir, "foo", "bar")
            os.makedirs(work_dir)
            env_file = os.path.join(tmpdir, "environment.yml")
            with open(env_file, "w") as f:
                f.write("\n")
            with working_directory(work_dir):
                self.assertEqual(
                    cenv_script.find_environment_file(), Path(env_file).resolve(),
                )

    def test_env_file_not_found(self):
        # env file must not be in tmp directory path or its ancestors!
        # otherwise test will not work as expected!
        with TemporaryDirectory() as tmpdir:
            with working_directory(tmpdir):
                with self.assertRaises(cenv_script.CondaEnvException):
                    cenv_script.find_environment_file()

    def test_command_line_interface(self):
        """Test the CLI."""
        runner = CliRunner()
        result = runner.invoke(cli.main)
        assert result.exit_code == 0
        help_result = runner.invoke(cli.main, ["--help"])
        assert help_result.exit_code == 0
        assert "--help" in help_result.output


cenv/cenv.py
"""Main module."""

import json
import os
import re
import shutil
import subprocess
import sys
from pathlib import Path
from typing import List, Optional

import yaml

ENV_FILE = "environment.yml"


class CondaEnvException(Exception):
    pass


def find_environment_file():
    p = Path(os.getcwd()).resolve()
    while True:
        env_file = p / ENV_FILE
        if env_file.is_file():
            return env_file
        if p.parents:
            p = p.parent
            continue
        raise CondaEnvException(
            "environment.yml file not find in '%s' or in any of parent directories"
            % os.getcwd()
        )


def get_conda():
    if sys.platform.startswith("win"):
        return "conda.bat"
    return "conda"


def print_args(args):
    def escape(arg):
        if arg.find(" ") > -1:
            return '"%s"' % arg
        return arg

    print(">>>", " ".join(map(escape, args)))


def in_directory(file_name, dir_name):
    return os.path.realpath(file_name).startswith(os.path.realpath(dir_name) + os.sep)


class CondaEnv:
    def __init__(self):
        super().__init__()
        self._conda = get_conda()
        self._env_file = find_environment_file()
        with open(self._env_file) as f:
            self._data = yaml.safe_load(f)
        data = subprocess.check_output([self._conda, "info", "--json"])
        data = json.loads(data)
        active_name = data["active_prefix_name"]
        active_prefix = data["active_prefix"]
        if active_name != self._data["name"]:
            raise CondaEnvException(
                f"Active environment is {active_name} but {ENV_FILE} points to {self._data['name']}"
            )
        if "prefix" in self._data and active_prefix != self._data["prefix"]:
            raise CondaEnvException(
                f"Active environment is located in {active_prefix} but {ENV_FILE} points to {self._data['prefix']}"
            )

        python_exe = shutil.which("python")
        if not python_exe:
            raise CondaEnvException("Python not found in path")
        # The following check is quite strict, but I think it is better to keep it. See comments below.
        if not in_directory(python_exe, active_prefix):
            raise CondaEnvException(
                f"Python '{python_exe}' is not in conda prefix '{active_prefix}'"
            )

    @staticmethod
    def pip_cmd(args):
        return [
            # disabled due to: https://github.com/conda/conda/issues/9572
            # "run", "-n", self._data["name"], "python",
            # This can lead to installing into the wrong place, but checks in the __init__ should help
            os.path.realpath(shutil.which("python")),
            "-m",
            "pip",
        ] + args

    def _exec_pip(self, args):
        args = self.pip_cmd(args)
        # return self._exec_conda(args)
        print_args(args)
        exit_code = subprocess.call(args)
        print("-" * 80)
        print("python -m pip finished with exit code: %d" % exit_code)
        return exit_code

    def _exec_conda(self, args):
        args = [self._conda] + args
        print_args(args)
        exit_code = subprocess.call(args)
        print("-" * 80)
        print("conda finished with exit code: %d" % exit_code)
        return exit_code

    @staticmethod
    def parse_pkg(pkg_spec: str):
        m = re.match(r"^(git|hg|svn|bzr)\+.*|^[\w-]+", pkg_spec)
        if m:
            return m.group(0)
        raise CondaEnvException("Failed to parse package specification '%s'" % pkg_spec)

    def _spec_add_package(self, deps: List[str], package: str) -> bool:
        """Add given package to a deps list if it is not already there

        :param deps: list of current dependencies
        :param package: package spec that should be added
        :return: True when deps list was mutated, False overwise
        """
        name = self.parse_pkg(package)
        for i, pkg in enumerate(deps):
            if not isinstance(pkg, str):
                continue
            pkg = pkg.strip()
            n = self.parse_pkg(pkg)
            if n == name:
                if pkg != package:
                    print(f"Updating spec from {pkg} to {package} ...")
                    deps[i] = package
                    break
                print(f"Same package spec already found: {pkg}")
                return False
        else:
            print(f"Adding package spec {package} to dependencies ...")
            deps.append(package)
        return True

    def install(self, package: str):
        package = package.strip()
        deps = self._get_deps()
        if not self._spec_add_package(deps, package):
            return

        exit_code = self._exec_conda(["install", "-n", self._data["name"], package])
        if exit_code != 0:
            raise CondaEnvException("Bad conda exitcode: %d" % exit_code)

        name = self.parse_pkg(package)
        if not self.check_installed(name):
            raise CondaEnvException(f"Package {name} was not installed")
        print("Verified that package has been installed")
        self._write_env_file()

    def check_installed(self, name):
        data = subprocess.check_output(
            [self._conda, "env", "export", "-n", self._data["name"]]
        )
        data = yaml.safe_load(data.decode("utf-8"))
        names = set(
            self.parse_pkg(x)
            for x in data.get("dependencies", [])
            if isinstance(x, str)
        )
        return name in names

    def pip_install(self, package: str):
        package = package.strip()
        deps = self._get_pip_deps()
        if not self._spec_add_package(deps, package):
            return

        exit_code = self._exec_pip(["install", package])
        if exit_code != 0:
            raise CondaEnvException("Bad conda+pip exitcode: %d" % exit_code)

        name = self.parse_pkg(package)
        if not self.check_pip_installed(name):
            raise CondaEnvException(
                f"Package {name} was not installed (not found in pip freeze)"
            )
        print("Verified that package has been installed")
        self._write_env_file()

    def check_pip_installed(self, name):
        data = subprocess.check_output(self.pip_cmd(["freeze"]))
        names = set(
            self.parse_pkg(l.strip()) for l in data.decode("utf-8").split("\n") if l
        )
        return name in names

    def _spec_rm_package(
        self, deps: List[str], package: str
    ) -> (Optional[str], List[str]):
        """Remove package from the deps list if it is present

        :param deps: current list of packages
        :param package: spec containing a package name that should be removed
        :return: tuple
          - package name if it was found or none
          - new list of packages
        """
        name = self.parse_pkg(package)
        new_deps = []
        to_remove = 0
        for pkg in deps:
            if not isinstance(pkg, str):
                continue
            n = self.parse_pkg(pkg)
            if n == name:
                to_remove += 1
                continue
            new_deps.append(pkg)

        if to_remove == 0:
            return None, new_deps

        if to_remove > 1:
            print("Warning: more than one spec matched")
        return name, new_deps

    def remove(self, package: str):
        package = package.strip()
        name, new_deps = self._spec_rm_package(self._get_deps(), package)
        self._set_deps(new_deps)
        if name is None:
            print("Specified package '%s' not found" % self.parse_pkg(package))
            return

        exit_code = self._exec_conda(["remove", "-n", self._data["name"], name])
        if exit_code != 0:
            raise CondaEnvException("Bad conda exitcode: %d" % exit_code)

        if self.check_installed(name):
            raise CondaEnvException(f"Package {name} was not removed")
        self._write_env_file()

    def pip_remove(self, package: str):
        package = package.strip()
        name, new_deps = self._spec_rm_package(self._get_pip_deps(), package)
        self._set_pip_deps(new_deps)
        if name is None:
            print(
                "Specified package '%s' not found in pip section"
                % self.parse_pkg(package)
            )
            return

        exit_code = self._exec_pip(["uninstall", name])
        if exit_code != 0:
            raise CondaEnvException("Bad conda exitcode: %d" % exit_code)

        if self.check_pip_installed(name):
            raise CondaEnvException(
                f"Package {name} was not removed (found in pip freeze)"
            )
        self._write_env_file()

    def _write_env_file(self):
        with open(self._env_file, "w") as f:
            yaml.dump(self._data, f)
        print("Updated %s" % ENV_FILE)

    def _get_deps(self):
        if "dependencies" not in self._data:
            self._data["dependencies"] = []
        return self._data["dependencies"]

    def _set_deps(self, value):
        self._data["dependencies"] = value

    def _get_pip_deps(self):
        for item in self._get_deps():
            if isinstance(item, dict) and "pip" in item:
                return item["pip"]
        self._data["dependencies"].append({"pip": []})
        return self._data["dependencies"][-1]["pip"]

    def _set_pip_deps(self, value):
        for item in self._get_deps():
            if isinstance(item, dict) and "pip" in item:
                item["pip"] = value
                return
        self._data["dependencies"].append({"pip": []})
        self._data["dependencies"][-1]["pip"] = value


cenv_script/cli.py
"""Console script for cenv_script."""
import sys
from functools import wraps
import traceback
import click
from cenv_script.cenv_script import CondaEnv, CondaEnvException


def catch(func):
    @wraps(func)
    def wrapper(ctx, *args, **kwargs):
        try:
            func(ctx, *args, **kwargs)
        except CondaEnvException as ex:
            if ctx.obj['verbose']:
                print("\n")
                traceback.print_exc()
                print("\n")
            print("error:", ex)
            return 1
        return 0

    return wrapper


@click.group()
@click.option("--verbose", is_flag=True, help="print exceptions traceback")
@click.pass_context
def main(ctx, verbose=False):
    ctx.ensure_object(dict)
    ctx.obj['verbose'] = verbose


@main.command()
@click.pass_context
@click.argument("package")
@catch
def install(ctx, package):
    """Install conda package"""
    e = CondaEnv()
    e.install(package)


@main.command()
@click.pass_context
@click.argument("package")
@catch
def remove(ctx, package):
    """Remove conda package"""
    e = CondaEnv()
    e.remove(package)


@main.command()
@click.pass_context
@click.argument("package")
@catch
def pip_install(ctx, package):
    """Install package with pip"""
    e = CondaEnv()
    e.pip_install(package)


@main.command()
@click.pass_context
@click.argument("package")
@catch
def pip_remove(ctx, package):
    """Remove package with pip"""
    e = CondaEnv()
    e.pip_remove(package)


if __name__ == "__main__":
    sys.exit(main(obj={}))  # pylint: disable-all # pragma: no cover


cenv_script/cenv_script.py
"""Main module."""

import json
import os
import re
import shutil
import subprocess
import sys
from pathlib import Path
from typing import List, Optional

import yaml

ENV_FILE = "environment.yml"


class CondaEnvException(Exception):
    pass


def find_environment_file():
    p = Path(os.getcwd()).resolve()
    while True:
        env_file = p / ENV_FILE
        if env_file.is_file():
            return env_file
        if p.parents:
            p = p.parent
            continue
        raise CondaEnvException(
            "environment.yml file not find in '%s' or in any of parent directories"
            % os.getcwd()
        )


def get_conda():
    if sys.platform.startswith("win"):
        return "conda.bat"
    return "conda"


def print_args(args):
    def escape(arg):
        if arg.find(" ") > -1:
            return '"%s"' % arg
        return arg

    print(">>>", " ".join(map(escape, args)))


def in_directory(file_name, dir_name):
    return os.path.realpath(file_name).startswith(os.path.realpath(dir_name) + os.sep)


class CondaEnv:
    def __init__(self):
        super().__init__()
        self._conda = get_conda()
        self._env_file = find_environment_file()
        with open(self._env_file) as f:
            self._data = yaml.safe_load(f)
        data = subprocess.check_output([self._conda, "info", "--json"])
        data = json.loads(data)
        active_name = data["active_prefix_name"]
        active_prefix = data["active_prefix"]
        if active_name != self._data["name"]:
            raise CondaEnvException(
                f"Active environment is {active_name} but {ENV_FILE} points to {self._data['name']}"
            )
        if "prefix" in self._data and active_prefix != self._data["prefix"]:
            raise CondaEnvException(
                f"Active environment is located in {active_prefix} but {ENV_FILE} points to {self._data['prefix']}"
            )

        python_exe = shutil.which("python")
        if not python_exe:
            raise CondaEnvException("Python not found in path")
        # The following check is quite strict, but I think it is better to keep it. See comments below.
        if not in_directory(python_exe, active_prefix):
            raise CondaEnvException(
                f"Python '{python_exe}' is not in conda prefix '{active_prefix}'"
            )

    @staticmethod
    def pip_cmd(args):
        return [
            # disabled due to: https://github.com/conda/conda/issues/9572
            # "run", "-n", self._data["name"], "python",
            # This can lead to installing into the wrong place, but checks in the __init__ should help
            os.path.realpath(shutil.which("python")),
            "-m",
            "pip",
        ] + args

    def _exec_pip(self, args):
        args = self.pip_cmd(args)
        # return self._exec_conda(args)
        print_args(args)
        exit_code = subprocess.call(args)
        print("-" * 80)
        print("python -m pip finished with exit code: %d" % exit_code)
        return exit_code

    def _exec_conda(self, args):
        args = [self._conda] + args
        print_args(args)
        exit_code = subprocess.call(args)
        print("-" * 80)
        print("conda finished with exit code: %d" % exit_code)
        return exit_code

    @staticmethod
    def parse_pkg(pkg_spec: str):
        m = re.match(r"^(git|hg|svn|bzr)\+.*|^[\w-]+", pkg_spec)
        if m:
            return m.group(0)
        raise CondaEnvException("Failed to parse package specification '%s'" % pkg_spec)

    def _spec_add_package(self, deps: List[str], package: str) -> bool:
        """Add given package to a deps list if it is not already there

        :param deps: list of current dependencies
        :param package: package spec that should be added
        :return: True when deps list was mutated, False overwise
        """
        name = self.parse_pkg(package)
        for i, pkg in enumerate(deps):
            if not isinstance(pkg, str):
                continue
            pkg = pkg.strip()
            n = self.parse_pkg(pkg)
            if n == name:
                if pkg != package:
                    print(f"Updating spec from {pkg} to {package} ...")
                    deps[i] = package
                    break
                print(f"Same package spec already found: {pkg}")
                return False
        else:
            print(f"Adding package spec {package} to dependencies ...")
            deps.append(package)
        return True

    def install(self, package: str):
        package = package.strip()
        deps = self._get_deps()
        if not self._spec_add_package(deps, package):
            return

        exit_code = self._exec_conda(["install", "-n", self._data["name"], package])
        if exit_code != 0:
            raise CondaEnvException("Bad conda exitcode: %d" % exit_code)

        name = self.parse_pkg(package)
        if not self.check_installed(name):
            raise CondaEnvException(f"Package {name} was not installed")
        print("Verified that package has been installed")
        self._write_env_file()

    def check_installed(self, name):
        data = subprocess.check_output(
            [self._conda, "env", "export", "-n", self._data["name"]]
        )
        data = yaml.safe_load(data.decode("utf-8"))
        names = set(
            self.parse_pkg(x)
            for x in data.get("dependencies", [])
            if isinstance(x, str)
        )
        return name in names

    def pip_install(self, package: str):
        package = package.strip()
        deps = self._get_pip_deps()
        if not self._spec_add_package(deps, package):
            return

        exit_code = self._exec_pip(["install", package])
        if exit_code != 0:
            raise CondaEnvException("Bad conda+pip exitcode: %d" % exit_code)

        name = self.parse_pkg(package)
        if not self.check_pip_installed(name):
            raise CondaEnvException(
                f"Package {name} was not installed (not found in pip freeze)"
            )
        print("Verified that package has been installed")
        self._write_env_file()

    def check_pip_installed(self, name):
        data = subprocess.check_output(self.pip_cmd(["freeze"]))
        names = set(
            self.parse_pkg(l.strip()) for l in data.decode("utf-8").split("\n") if l
        )
        return name in names

    def _spec_rm_package(
        self, deps: List[str], package: str
    ) -> (Optional[str], List[str]):
        """Remove package from the deps list if it is present

        :param deps: current list of packages
        :param package: spec containing a package name that should be removed
        :return: tuple
          - package name if it was found or none
          - new list of packages
        """
        name = self.parse_pkg(package)
        new_deps = []
        to_remove = 0
        for pkg in deps:
            if not isinstance(pkg, str):
                continue
            n = self.parse_pkg(pkg)
            if n == name:
                to_remove += 1
                continue
            new_deps.append(pkg)

        if to_remove == 0:
            return None, new_deps

        if to_remove > 1:
            print("Warning: more than one spec matched")
        return name, new_deps

    def remove(self, package: str):
        package = package.strip()
        name, new_deps = self._spec_rm_package(self._get_deps(), package)
        self._set_deps(new_deps)
        if name is None:
            print("Specified package '%s' not found" % self.parse_pkg(package))
            return

        exit_code = self._exec_conda(["remove", "-n", self._data["name"], name])
        if exit_code != 0:
            raise CondaEnvException("Bad conda exitcode: %d" % exit_code)

        if self.check_installed(name):
            raise CondaEnvException(f"Package {name} was not removed")
        self._write_env_file()

    def pip_remove(self, package: str):
        package = package.strip()
        name, new_deps = self._spec_rm_package(self._get_pip_deps(), package)
        self._set_pip_deps(new_deps)
        if name is None:
            print(
                "Specified package '%s' not found in pip section"
                % self.parse_pkg(package)
            )
            return

        exit_code = self._exec_pip(["uninstall", name])
        if exit_code != 0:
            raise CondaEnvException("Bad conda exitcode: %d" % exit_code)

        if self.check_pip_installed(name):
            raise CondaEnvException(
                f"Package {name} was not removed (found in pip freeze)"
            )
        self._write_env_file()

    def _write_env_file(self):
        with open(self._env_file, "w") as f:
            yaml.dump(self._data, f, sort_keys=False)
        print("Updated %s" % ENV_FILE)

    def _get_deps(self):
        if "dependencies" not in self._data:
            self._data["dependencies"] = []
        return self._data["dependencies"]

    def _set_deps(self, value):
        self._data["dependencies"] = value

    def _get_pip_deps(self):
        for item in self._get_deps():
            if isinstance(item, dict) and "pip" in item:
                return item["pip"]
        self._data["dependencies"].append({"pip": []})
        return self._data["dependencies"][-1]["pip"]

    def _set_pip_deps(self, value):
        for item in self._get_deps():
            if isinstance(item, dict) and "pip" in item:
                item["pip"] = value
                return
        self._data["dependencies"].append({"pip": []})
        self._data["dependencies"][-1]["pip"] = value
